Support content trust offline

dmcgowan opened the pull request:
Ignore ping errors in notary repository setup since Notary is capable of acting in offline mode, making use of cache TUF data. When ping is not successful, notary should still be attempted without error.

endophage commented:
On the off chance it's a temporary or one off interruption, is it possible to make a test that fails the initial ping but gets a valid 401 with challenge response during a later request? Just wondering what happens in that situation.

Otherwise, LGTM.

thaJeztah commented:
needs a rebase, @dmcgowan

cpuguy83 commented:
LGTM

jfrazelle commented:
is this fixing a bug?

endophage commented:
@jfrazelle yes. Currently any operation that has content trust enabled will fail when you're offline. It should attempt to use local cache, for which normal TUF expiry rules will apply.

jfrazelle commented:
ok

diogomonica commented on api/client/trust.go:
```diff
@@ -248,6 +252,8 @@ func notaryError(err error) error {
 		return fmt.Errorf("remote repository out-of-date: %v", err)
 	case trustmanager.ErrKeyNotFound:
 		return fmt.Errorf("signing keys not found: %v", err)
+	case *net.OpError:
```
Is `notaryError()` used anywhere?

dmcgowan replied:
yes, every time we call notary we wrap the error in this call

diogomonica replied:
a search for `notaryError` on docker/docker only brings up this result. What am I missing?

cpuguy83 replied:
<img width="517" alt="screen shot 2015-09-03 at 10 22 18 pm" src="https://cloud.githubusercontent.com/assets/799078/9675357/627581b0-528a-11e5-8a4c-7a1b9f5cf657.png">

diogomonica replied:
Yeah, just grepped and got the same. Somehow github's code search isn't returning it.

diogomonica commented:
LGTM, but when reviewing this I noticed that we don't have a test for either a `docker pull` or a `docker run` with an offline notary-server and expired timestamp (> 3 weeks in the future).

jfrazelle commented on api/client/trust.go:
```diff
@@ -248,6 +252,8 @@ func notaryError(err error) error {
 	case trustmanager.ErrKeyNotFound:
 		return fmt.Errorf("signing keys not found: %v", err)
+	case *net.OpError:
+		return fmt.Errorf("error contacting notary server: %s", err)
```
can this be a %v, the others are, and i know its a nit pick buttttt <3

diogomonica replied:
jess is right. this is the right thing. we must do this.

jfrazelle replied:
lol

On Thu, Sep 3, 2015 at 6:03 PM, Diogo Mónica <notifications@github.com>
wrote:

> In api/client/trust.go
> <https://github.com/docker/docker/pull/15864#discussion_r38713553>:
>
> > @@ -248,6 +252,8 @@ func notaryError(err error) error {
> >  		return fmt.Errorf("remote repository out-of-date: %v", err)
> >  	case trustmanager.ErrKeyNotFound:
> >  		return fmt.Errorf("signing keys not found: %v", err)
> > +	case *net.OpError:
> > +		return fmt.Errorf("error contacting notary server: %s", err)
>
> jess is right. this is the right thing. we must do this.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/15864/files#r38713553>.
>

tiborvass commented on api/client/trust.go:
```diff
@@ -144,15 +144,19 @@ func (cli *DockerCli) getNotaryRepository(repoInfo *registry.RepositoryInfo, aut
-	challengeManager := auth.NewSimpleChallengeManager()
-	if err := challengeManager.AddResponse(resp); err != nil {
-		return nil, err
+		if err := challengeManager.AddResponse(resp); err != nil {
```
@dmcgowan can you explain what this does with a comment?

dmcgowan committed:
Ignore ping errors in notary repository setup

Notary is capable of acting in offline mode, making use of cache TUF data.
When ping is not successful, notary should still be attempted without error.

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

aaronlehmann commented:
Needs another rebase.

RichardScothern commented:
LGTM.  (but still needs a rebase)

calavera commented:
LGTM
